### PR TITLE
BlueskyRun loading time

### PR DESCRIPTION
**Pull request by gwbischof**

### CODE REVIEW

File: xicam/gui/bluesky/search.py [hunk 1]
```diff
@@ -176,14 +183,16 @@ def set_selected_catalog(self, item):
...
 
+    @timeit
     def check_for_new_entries(self):
         # check for any new results and add them to the queue for later processing
-        for uid, entry in itertools.islice(self._results_catalog.items(), MAX_SEARCH_RESULTS):
-            if uid in self._results:
-                continue
-            self._results.append(uid)
-            self._new_entries.put(entry)
+        for uid in list(self._results_catalog)[:MAX_SEARCH_RESULTS]:
```
Comment (gwbischof) [inline xicam/gui/bluesky/search.py]: This change helped the execution time of this function for reloads ~3s -> 1ms for a query with about 10 runs, but didn't help the initial query much

File: xicam/gui/bluesky/search.py [hunk 2]
```diff
@@ -47,11 +47,19 @@
...
+        return ret
+    return wrap
+
 
 def default_search_result_row(entry):
-    metadata = entry.describe()['metadata']
-    start = metadata['start']
-    stop = metadata['stop']
+    start = entry.metadata['start']
+    stop = entry.metadata['stop']
```
Comment (gwbischof) [inline xicam/gui/bluesky/search.py:53]: This changed help the execution time of show_results. For a set of about 10 runs it took about 2.2s before and ~1ms after. describe() calls BlueskyRun._load(), which is not needed here because the run has already been loaded. BlueskyRun._load() takes a while to execute, so I tried to minimize the number of times it is called.

**Review by gwbischof — COMMENTED** [review]
I think that this is all we can do on the Xicam side. BlueskyRun._load() is only called once for each item now, and refreshes don't call load for all items.  I think we need to investigate BlueskyRun._load() to improve this more.

**Comment by gwbischof** [comment]
BlueskyRun._load() takes ~180ms per call. That accounts for almost of the time it takes to load the Bluesky runs.

**Comment by gwbischof** [comment]
I noticed that before these changes that every RELOAD_INTERVAL, show results was called and was taking 3s each time for the set of ~10runs. With these changes the reloads only takes  ~1ms.

File: xicam/gui/bluesky/search.py (hunk 1 above)
Comment (danielballan) [reply xicam/gui/bluesky/search.py]: This will make the performance much worse for large results sets because it will pull the entire results set into memory, rather than just the first `MAX_SEARCH_RESULTS` entries. I don't think we can do this.

**Review by danielballan — CHANGES_REQUESTED** [review]

File: xicam/gui/bluesky/search.py (hunk 2 above)
Comment (danielballan) [reply xicam/gui/bluesky/search.py:53]: Now that we have a solution for the serialization issue, is this change still impactful? I recall that we adopted `describe()` intentionally here, though I'm unfortunately fuzzy on the reasoning.

File: xicam/gui/bluesky/search.py (hunk 1 above)
Comment (danielballan) [reply xicam/gui/bluesky/search.py]: Thinking on this some, I'm completely stumped as to where that speedup would come from. In this toy example, you can see that `list(...)[:N]` accesses _every_ element, including the ones past `N` that aren't needed, whereas `itertools.islice(..., N)` accesses just the elements we need.

```py
In [8]: def my_iterator(): 
   ...:     for i in range(10): 
   ...:         print(f'accessing item {i}') 
   ...:         yield i 
   ...:                                                                                                                                                                                       

In [9]: for _ in list(my_iterator())[:5]: 
   ...:     ... 
   ...:                                                                                                                                                                                       
accessing item 0
accessing item 1
accessing item 2
accessing item 3
accessing item 4
accessing item 5
accessing item 6
accessing item 7
accessing item 8
accessing item 9

In [10]: import itertools                                                                                                                                                                     

In [11]: for _ in itertools.islice(my_iterator(), 5): 
    ...:     ... 
    ...:                                                                                                                                                                                      
accessing item 0
accessing item 1
accessing item 2
accessing item 3
accessing item 4
```

File: xicam/gui/bluesky/search.py [hunk 3]
```diff
@@ -82,7 +90,7 @@ def __init__(self, catalog):
         self.catalog_selection_model = CatalogSelectionModel()
         self.search_results_model = SearchResultsModel(self)
         self._subcatalogs = []  # to support lookup by item's positional index
-        self._results = []  # to support lookup by item's positional index
+        self._results = set()  # to support lookup by item's positional index
```
Comment (dylanmcreynolds) [inline xicam/gui/bluesky/search.py]: Why the change to a set?

**Review by danielballan — COMMENTED** [review]
I like use of `Qt.UserRole`  to store the uid next to the data rather than keeping a separate list (`_results`). I also like the pattern of using a set to keep track of what we have already thrown into the queue. I'm glad we get to keep the speed-up due to using `set` while avoiding parallel data structures.

I am still unsure about the changes to `default_search_result_row`. A `LocalCatalogEntry` has no `metadata` attribute; accessing `entry.metadata` triggers intake's possibly-too-magical fallback mechanism where the `AttributeError` raised by `entry.metadata` is caught and `entry._get_default_source().metadata` is evaluated and returned. Thus, I don't see how stepping around `describe()` actually saves any work.

**Comment by gwbischof** [comment]
`entry` is not a `LocalCatalogEntry`, it is a `BlueskyRun`
